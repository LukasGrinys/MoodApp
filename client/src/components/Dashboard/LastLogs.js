import React from 'react';
import { Link } from 'react-router-dom';
import parentStyles from './Dashboard.module.scss';
import styles from './LastLogs.module.scss';
import LoadingNetItem from '../Loading/loadingNetItem';
import { useTheme } from '../../contexts/ThemeContext';
import { useSelector } from 'react-redux';
import Button from '../Button/Button';
import classNames from 'classnames';

const LastLogs = () => {
    const darkTheme = useTheme();
    const { lastLogs, isFetchingLogs } = useSelector( ({logs}) => {
        return {
            lastLogs : logs.lastLogs,
            isFetchingLogs : logs.isFetchingLogs
        }
    });

    return (
        <div className={parentStyles.gridItem}>
            <h3 className={parentStyles.gridItemHeader}>Last logs</h3>
            {
                isFetchingLogs ? (
                    <LoadingNetItem/>
                ) : ( (!lastLogs || !lastLogs.length) ? (
                        <div className={styles.greyText}>
                            There are no logs at the moment <br/><br/>
                            Write Your first one, by pressing the green button at the top
                        </div> 
                    ) : (
                        <div className={styles.logSection}> 
                            <div className={styles.logContainer}>
                                { lastLogs.slice(0,3).map( (log, index) => {
                                    return (
                                        <div key={index} className={styles.logItem}>
                                            <div className={styles.logContent}>
                                                <div className={classNames(
                                                    styles.ratingBox,
                                                    darkTheme && styles.dark
                                                )}>
                                                    {log.rating}
                                                </div>
                                                <div className={styles.rightSection}>
                                                    <div className={styles.date}>{log.date}</div>
                                                    <div className={styles.text}>{log.text}</div>
                                                </div>
                                            </div>
                                            <Link 
                                                to={`/logs/${log.id}`}
                                                className={classNames(
                                                    styles.link,
                                                    darkTheme && styles.dark
                                                )}
                                            >
                                                Read more
                                            </Link> 
                                        </div>
                                    )
                                })}
                            </div>
                            <Link to="/logs">
                                <Button>All Logs</Button>
                            </Link>
                        </div> )
                )
            }
            
        </div>
    )
};

export default LastLogs;