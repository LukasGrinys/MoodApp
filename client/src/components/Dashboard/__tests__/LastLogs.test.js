import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import LastLogs from '../LastLogs';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import LoadingNetItem from '../../Loading/loadingNetItem';

const mockStore = configureStore();
const noLogsText = 'There are no logs at the moment Write Your first one, by pressing the green button at the top';
const testLog = {
    id: '001',
    rating: 10,
    date: "TEST-DATE",
    text: "Test text"
}

describe('Last logs item: loading', () => {
    const state = {
        user: {},
        logs: {
            lastLogs : {
                isFetching: true,
                logs: null
            }
        }
    }

    const store = mockStore(state);

    const wrapper = mount(
        <Provider store={store}>
            <LastLogs/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has LoadingNetItem widget', () => {
        expect(wrapper.find(LoadingNetItem).length).toBe(1);
    });
});

describe('Last logs item: has no logs', () => {
    const state = {
        user: {},
        logs: {
            lastLogs : {
                isFetching: false,
                logs: []
            }
        }
    }
    const store = mockStore(state);

    const wrapper = mount(
        <Provider store={store}>
            <LastLogs/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Shows text with "There are no logs"', () => {
        expect(wrapper.find('.greyText').text()).toBe(noLogsText);
    });
});


describe('Last logs item: has logs', () => {
    const state = {
        user: {},
        logs: {
            lastLogs : {
                isFetching: false,
                logs: [ testLog ]
            }
        }
    }
    const store = mockStore(state);

    const wrapper = mount(
        <Router>
            <Provider store={store}>
                <LastLogs/>
            </Provider>
        </Router>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Shows log rating box with correct rating', () => {
        expect(wrapper.find('.ratingBox').text()).toBe(testLog.rating.toString());
    });

    test('Shows log date', () => {
        expect(wrapper.find('.date').text()).toBe(testLog.date);
    });

    test('Shows log text', () => {
        expect(wrapper.find('.text').text()).toBe(testLog.text);
    });

    test('Has rendered a link for log reading', () => {
        expect(wrapper.find('.link').at(0).props()).toEqual({
            to: `/logs/${testLog.id}`,
            className: 'link',
            children: 'Read more'
        });
    });

    test('Has rendered as much items as there are', () => {
        expect(wrapper.find('.logItem').length).toBe(state.logs.lastLogs.logs.length);
    });

    test('Has a button for more logs', () => {
        expect(wrapper.find('button').length).toBe(1);
        expect(wrapper.find('button').text()).toBe('All Logs');
    });
});

const multipleTestLogs = [
    {
        id: '001',
        rating: 10,
        date: "TEST-DATE",
        text: "Test text"
    },
    {
        id: '002',
        rating: 9,
        date: "TEST-DATE",
        text: "Test text"
    },
    {
        id: '003',
        rating: 8,
        date: "TEST-DATE",
        text: "Test text"
    },
    {
        id: '004',
        rating: 7,
        date: "TEST-DATE",
        text: "Test text"
    }
]

test('If there is more than 3 logs, it shows only the last 3', () => {
    const state = {
        user: {},
        logs: {
            lastLogs : {
                isFetching: false,
                logs: multipleTestLogs
            }
        }
    }
    const store = mockStore(state);

    const wrapper = mount(
        <Router>
            <Provider store={store}>
                <LastLogs/>
            </Provider>
        </Router>
    );

    expect(wrapper.find('.logItem').length).toBe(3);
})