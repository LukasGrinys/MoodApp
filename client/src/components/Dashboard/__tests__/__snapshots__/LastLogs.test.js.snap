// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Last logs item: has logs Renders properly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <LastLogs>
        <div
          className="gridItem"
        >
          <h3
            className="gridItemHeader"
          >
            Last logs
          </h3>
          <div
            className="logSection gridItemContent"
          >
            <div
              className="logContainer"
            >
              <div
                className="logItem"
                key="0"
              >
                <div
                  className="logContent"
                >
                  <div
                    className="ratingBox"
                  >
                    10
                  </div>
                  <div
                    className="rightSection"
                  >
                    <div
                      className="date"
                    >
                      TEST-DATE
                    </div>
                    <div
                      className="text"
                    >
                      Test text
                    </div>
                  </div>
                </div>
                <Link
                  className="link"
                  to="/logs/001"
                >
                  <LinkAnchor
                    className="link"
                    href="/logs/001"
                    navigate={[Function]}
                  >
                    <a
                      className="link"
                      href="/logs/001"
                      onClick={[Function]}
                    >
                      Read more
                    </a>
                  </LinkAnchor>
                </Link>
              </div>
            </div>
            <Link
              to="/logs"
            >
              <LinkAnchor
                href="/logs"
                navigate={[Function]}
              >
                <a
                  href="/logs"
                  onClick={[Function]}
                >
                  <Button>
                    <button
                      className="button btnPrimary"
                      onClick={[Function]}
                      type="button"
                    >
                      All Logs
                    </button>
                  </Button>
                </a>
              </LinkAnchor>
            </Link>
          </div>
        </div>
      </LastLogs>
    </Provider>
  </Router>
</BrowserRouter>
`;

exports[`Last logs item: has no logs Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <LastLogs>
    <div
      className="gridItem"
    >
      <h3
        className="gridItemHeader"
      >
        Last logs
      </h3>
      <div
        className="greyText"
      >
        There are no logs at the moment 
        <br />
        <br />
        Write Your first one, by pressing the green button at the top
      </div>
    </div>
  </LastLogs>
</Provider>
`;

exports[`Last logs item: loading Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <LastLogs>
    <div
      className="gridItem"
    >
      <h3
        className="gridItemHeader"
      >
        Last logs
      </h3>
      <LoadingNetItem>
        <div
          className="loadingScreen"
        >
          <div
            className="barContainer"
          >
            <div
              className="bar"
            />
            <div
              className="bar delay1"
            />
            <div
              className="bar delay2"
            />
          </div>
        </div>
      </LoadingNetItem>
    </div>
  </LastLogs>
</Provider>
`;
